abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1807 @@
+  allow the host to finalize and delete that context.
+
+
+### `proxy_on_log`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - none
+
+Called after the host is done processing context, but before
+releasing its state.
+
+This callback can be used for generating final log entries.
+
+
+### `proxy_on_delete`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - none
+
+Called when the host removes the context (`context_id`).
+
+This is used to signal that the plugin should stop tracking active
+context and remove all associated state.
+
+
+### `proxy_done`
+
+* params:
+  - none
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Indicates to the host that the plugin is done processing active
+context.
+
+This should be used after returning `false` in [`proxy_on_done`].
+
+Returned `status` value is:
+- `OK` on success.
+- `NOT_FOUND` when active context was not pending finalization.
+
+
+### `proxy_set_effective_context`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Changes the effective context to `context_id`.
+
+This can be used to change active context, e.g. during
+[`proxy_on_http_call_response`], [`proxy_on_grpc_receive`]
+and/or [`proxy_on_queue_ready`] callbacks.
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for unknown `context_id`.
+
+
+## Configuration
+
+### `proxy_on_vm_start`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (size_t) vm_configuration_size`
+* returns:
+  - `i32 (bool) status`
+
+Called when the host starts the WebAssembly Virtual Machine.
+
+Its configuration (of `vm_configuration_size`) can be retrieved using
+[`proxy_get_buffer_bytes`] with `buffer_id` set to `VM_CONFIGURATION`.
+
+Plugin must return one of the following values:
+- `true` to indicate that the configuration was processed successfully.
+- `false` to indicate that the configuration processing failed.
+
+
+### `proxy_on_configure`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (size_t) plugin_configuration_size`
+* returns:
+  - `i32 (bool) status`
+
+Called when the host starts the Proxy-Wasm plugin.
+
+Its configuration (of `plugin_configuration_size`) can be retrieved
+using [`proxy_get_buffer_bytes`] with `buffer_id` set to
+`PLUGIN_CONFIGURATION`.
+
+Plugin must return one of the following values:
+- `true` to indicate that the configuration was processed successfully.
+- `false` to indicate that the configuration processing failed.
+
+
+## Logging
+
+### `proxy_log`
+
+* params:
+  - `i32 (`[`proxy_log_level_t`]`) log_level`
+  - `i32 (const char *) message_data`
+  - `i32 (size_t) message_size`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Logs `message` at the `log_level`.
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for unknown `log_level`.
+- `INVALID_MEMORY_ACCESS` when `message_data` and/or `message_size`
+  point to invalid memory address.
+
+
+### `wasi_snapshot_preview1.fd_write`
+
```

Comment:
same.. summary comment if this is required, which would be odd. in fact, _start and this may be better put into an "optional support" section unless they are required.

Reply:
in general, it would be better to keep all things like this (things that are not defined in this ABI) as a historical appendix. That envoy's host exports a few functions is interesting info, so is that some compilers only have a wasi target. However, even though these topics affect portability, these  are different than the ABI in general which is used by hosts besides envoy such as mosn cc @antjack

Reply:
But those are neither historical not optional, and intercepting `fd_write` is actually quite useful when integrating with existing 3rd party libraries that print to `stdout` and/or `stderr`.

Ideally, all Proxy-Wasm plugins should use proper logging facilitates and use `proxy_log`, but that's not always possible.